[MRG] Fix LocalOutlierFactor's output for data with duplicated samples

HenriqueProj opened the pull request:
#### Reference Issues/PRs
Fixes #27839

#### What does this implement/fix? Explain your changes.
Previously, when the dataset had values repeat more times than the algorithm's number of neighbors, it miscalculates the outliers.
Because the distance between the duplicated samples is 0, the local reachability density is equal to 1e10. This leads to values that are close to the duplicated values having a really low `negative_outlier_factor_` (under -1e7), labeling them as outliers.

This fix checks if the minimum `negative_outlier_factor_` is under -1e7 and, if so, raises the number of neighbors to the number of occurrences of the most frequent value + 1, also raising a warning.

Notes: Added a `handle_duplicates` variable, which allows developers to manually handle the duplicate values, if desired.
Also added a `memory_limit` variable to avoid creating memory errors for really large datasets, which can also be changed manually by developers.

#### Any other comments?

HenriqueProj committed:
Fix scikit-learn#27839: Adjust LocalOutlierFactor for data with duplicated samples

Previously, when the dataset had values repeat more times than the algorithm's number of neighbors, it miscalculates the outliers.
Because the distance between the duplicated samples is 0, the local reachability density is equal to 1e10. This leads to values that are close to the duplicated values having a really low negative outlier factor (under -1e7), labeling them as outliers.
This fix checks if the minimum negative outlier factor is under -1e7 and, if so, raises the number of neighbors to the number of occurrences of the most frequent value + 1, also raising a warning.
Notes: Added a handle_duplicates variable, which allows developers to manually handle the duplicate values, if desired. Also added a memory_limit variable to avoid creating memory errors for really large datasets, which can also be changed manually by developers.

HenriqueProj committed:
Redo checks: Codecov server error

HenriqueProj committed:
Update changelog and update tests for 100% test coverage

HenriqueProj committed:
Fix changelog error

ogrisel commented:
I don't think I like the recursive automatic change to neighbors.

Maybe we should instead just warn the user when we detect the problem with very negative outlier factor values and let the user re-fit the model with a larger value of `n_neighbors` by themselves.

HenriqueProj committed:
Fix: Changed approach according to review
Removed automatic change to neighbors number and changed the warning
Also changed the associated test, to catch the warning.

HenriqueProj commented:
@ogrisel Changed the code. Now it only raises a warning, as suggested.

betatim commented on sklearn/neighbors/_lof.py:
```diff
@@ -317,6 +317,16 @@ def fit(self, X, y=None):
                 self.negative_outlier_factor_, 100.0 * self.contamination
             )
 
+        """
+        Verify if negative_outlier_factor_ values are within acceptable range
+        Novelty must also be false to detect outliers
+        """
```
```suggestion
        # Verify if negative_outlier_factor_ values are within acceptable range.
        # Novelty must also be false to detect outliers
```
No need for syntax abuse, use `#` instead :D

HenriqueProj replied:
Thank you for the feedback!

HenriqueProj committed:
Update sklearn/neighbors/_lof.py

Changed comment according to review

Co-authored-by: Tim Head <betatim@gmail.com>

HenriqueProj committed:
Merge branch 'main' into fix_lof_duplicate_samples

HenriqueProj committed:
Merge branch 'main' into fix_lof_duplicate_samples

HenriqueProj commented:
@betatim Is there anything more I can do?
Thank you.

OmarManzoor commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -359,3 +359,37 @@ def test_lof_dtype_equivalence(algorithm, novelty, contamination):
+    Check that LocalOutlierFactor raises a warning when duplicate values
+    in the training data are causing innacurate results.
+
+    Test for: https://github.com/scikit-learn/scikit-learn/issues/27839
```
```suggestion
    Non-regression test for:
    https://github.com/scikit-learn/scikit-learn/issues/27839
```

OmarManzoor reviewed (commented):
Thanks for the PR @HenriqueProj. I think this may require a changelog entry as a Fix.

HenriqueProj committed:
Update sklearn/neighbors/tests/test_lof.py

Changed test description according to review

Co-authored-by: Omar Salman <omar.salman@arbisoft.com>

HenriqueProj committed:
Merge branch 'main' into fix_lof_duplicate_samples

HenriqueProj committed:
Added changelog entry.

HenriqueProj commented:
@OmarManzoor Made the changes, is everything ok? Thanks for the feedback!

OmarManzoor commented on doc/whats_new/v1.6.rst:
```diff
@@ -80,14 +80,21 @@ Changelog
+:mod:`sklearn.neighbors`
+........................
+
+- |Fix| Added a warning in :class:`_lof.LocalOutlierFactor`'s `fit` method, which is
+  raised when duplicate values in the training data lead to inaccurate outlier detection.
+  :pr:`28773` by :user:`Henrique Caroço <HenriqueProj>`.
```
```suggestion
- |Fix| :class:`neighbors.LocalOutlierFactor` raises a warning in the `fit` method
   when duplicate values in the training data lead to inaccurate outlier detection.
  :pr:`28773` by :user:`Henrique Caroço <HenriqueProj>`.
```

OmarManzoor reviewed (commented):
Minor suggestions and also the conflict needs to be resolved.

HenriqueProj committed:
Merge branch 'main' into fix_lof_duplicate_samples

HenriqueProj committed:
Update doc/whats_new/v1.6.rst

Co-authored-by: Omar Salman <omar.salman@arbisoft.com>

HenriqueProj committed:
Update sklearn/neighbors/tests/test_lof.py

Co-authored-by: Omar Salman <omar.salman@arbisoft.com>

HenriqueProj committed:
Fixed documentation

HenriqueProj committed:
Merge branch 'main' into fix_lof_duplicate_samples

OmarManzoor commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -359,3 +359,38 @@ def test_lof_dtype_equivalence(algorithm, novelty, contamination):
+def test_lof_duplicate_samples():
+    """
+    Check that LocalOutlierFactor raises a warning when duplicate values
+    in the training data are cause inaccurate results.
```
```suggestion
    in the training data cause inaccurate results.
```

OmarManzoor commented on sklearn/neighbors/tests/test_lof.py:
```diff
@@ -359,3 +359,38 @@ def test_lof_dtype_equivalence(algorithm, novelty, contamination):
+
+    rng = np.random.default_rng(0)
+
+    # 100 times the number of elements of the example shown in the issue
```
```suggestion
```

OmarManzoor reviewed (approved):
LGTM. Thanks @HenriqueProj

OmarManzoor committed:
Update sklearn/neighbors/tests/test_lof.py

OmarManzoor committed:
Update sklearn/neighbors/tests/test_lof.py

HenriqueProj commented:
Thank you for all the feedback!
